module/dkg/broker.go

```diff
@@ -120,22 +120,29 @@ func (b *Broker) Broadcast(data []byte) {
 	}
 	maxedExpRetry := retry.WithMaxRetries(retryMax, expRetry)
 
-	f := func() {
+	b.unit.Launch(func() {
+		attempts := 1
 		err = retry.Do(context.Background(), maxedExpRetry, func(ctx context.Context) error {
 			err := b.dkgContractClient.Broadcast(bcastMsg)
 			if err != nil {
-				b.log.Error().Err(err).Msg("error broadcasting, retrying")
+				b.log.Error().Err(err).Msgf("error broadcasting, retrying (%x)", attempts)
 			}
+
+			attempts++
 			return retry.RetryableError(err)
 		})
 
+		// Various network can conditions can result in errors while broadcasting DKG messages,
+		// because failure to send an individual DKG message doesn't necessarily result in local or global DKG failure
+		// it is acceptable to log the error and move on.
 		if err != nil {
 			b.log.Error().Err(err).Msg("failed to broadcast message")
 		}
-		b.broadcasts++
-	}
 
-	b.unit.Launch(f)
+		b.unit.Lock()
+		b.broadcasts++
+		b.unit.Unlock()
```

Comment:
checking the broadcast count and incrementing it should happen in the same lock context, and before sending the broadcast message. Otherwise we have a race condition where two threads check the count, see 0, _don't_ log a warning, then both broadcast and increment the counter. We have broadcast twice but won't know about it.

Reply:
@kc1116 We still have this problem with the most recent change. 2 (or more) threads can:
* acquire the lock
* read `b.broadcasts`
* release the lock
before either thread completes the broadcast and increments `b.broadcasts` (because the read and the write are happening after 2 separate instances of acquiring the lock -- they are not occurring in the same lock context).

Instead we should:
```go
// acquire the lock
// read the broadcasts value and log accordingly
// increment the broadcasts value
// release the lock
// perform the actual broadcast
```

Reply:
Okay is it correct to increment broadcast even if the broadcasting fails ?

Reply:
> is it correct to increment broadcast even if the broadcasting fails ?

It is. We're counting the number of times the underlying DKG _requested_ a broadcast so we can detect an unhappy path.  I can see why there'd be confusion about that though -- do you mind adding a comment to that effect when we increment `broadcasts`?